src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -381,115 +353,131 @@ public override int Read(byte[] array, int offset, int count)
+                eventsToPoll |= Interop.Sys.PollEvents.POLLIN;
+            }
 
-                    if ((events & (Interop.Sys.PollEvents.POLLERR | Interop.Sys.PollEvents.POLLNVAL)) != 0)
-                    {
-                        throw new IOException();
-                    }
+            if (pollWriteEvents)
+            {
+                eventsToPoll |= Interop.Sys.PollEvents.POLLOUT;
+            }
 
-                    if ( (events & Interop.Sys.PollEvents.POLLOUT) == 0)
-                    {
-                        throw new TimeoutException(SR.Write_timed_out);
-                    }
-                }
+            Interop.Sys.PollEvents events = Interop.Sys.PollEvents.POLLNONE;
+            Interop.Sys.Poll(_handle,
+                             eventsToPoll,
+                             timeout,
+                             out events);
+            return events;
+        }
 
-                fixed (byte* bufPtr = array)
-                {
-                    numBytes = Interop.Sys.Write(_handle, bufPtr + offset, count);
-                }
+        internal void Write(byte[] array, int offset, int count, int timeout)
+        {
+            using (CancellationTokenSource cts = GetCancellationTokenSourceFromTimeout(timeout))
+            {
+                Task t = WriteAsync(array, offset, count, cts?.Token ?? CancellationToken.None);
 
-                if (numBytes == -1)
+                try
                 {
-                    throw new IOException();
+                    t.GetAwaiter().GetResult();
                 }
-
-                if (numBytes == 0)
+                catch (OperationCanceledException)
                 {
-                    throw new TimeoutException(SR.Write_timed_out);
+                    throw new TimeoutException();
                 }
-                count -= numBytes;
-                offset += numBytes;
+            }
+        }
+
+        public override IAsyncResult BeginWrite(byte[] array, int offset, int count, AsyncCallback userCallback, object stateObject)
+        {
+            return TaskToApm.Begin(WriteAsync(array, offset, count), userCallback, stateObject);
+        }
+
+        public override void EndWrite(IAsyncResult asyncResult)
+            => EndReadWrite(asyncResult);
+
+        private int EndReadWrite(IAsyncResult asyncResult)
+        {
+            try
+            {
+                return TaskToApm.End<int>(asyncResult);
+            }
+            catch (OperationCanceledException)
+            {
+                throw new TimeoutException();
             }
         }
 
         internal SafeFileHandle OpenPort(string portName)
         {
             SafeFileHandle handle = Interop.Serial.SerialPortOpen(portName);
+            if (handle.IsInvalid)
+            {
+                throw new UnauthorizedAccessException(string.Format(SR.UnauthorizedAccess_IODenied_Port, portName));
```

Comment:
The only reason this might fail is because of unauthorized access?  Is that the only exception we might throw on Windows?

Reply:
Good question. Windows does GetExceptionForLastWin32Error which in the most common case throws UnauthorizedAccessException. On Linux equivalent function (`Interop.GetIOException(Interop.Sys.GetLastErrorInfo())`) throws:
`System.IO.IOException: Device or resource busy`. The tests and users I think will expect UnauthorizedAccessException (even though IOException probably makes would make more sense if we rewrote it from scratch).

Is there any specific error you have in mind? I think always throwing UnauthorizedAccessException makes sense as it is the most common case but perhaps we could fix that in Interop.GetIOException.

thoughts?